src/graph.rs

```diff
@@ -17,15 +17,52 @@
 //!
 //! [`Arena`]: ../arena/struct.Arena.html
 //! [`StackGraph`]: struct.StackGraph.html
+//!
+//! There are several different kinds of node that can appear in a stack graph.  As we search for a
+//! path representing a name binding, each kind of node has different rules for how it interacts
+//! with the symbol and scope stacks:
+//!
+//!   - the singleton [_root node_][`RootNode`], which allows name binding paths to cross between
+//!     files
+//!   - [_exported_][`ExportedScopeNode`] and [_internal_][`InternalScopeNode`] _scopes_, which
+//!     define the name binding structure within a single file
+//!   - [_push symbol_][`PushSymbolNode`] and [_push scoped symbol_][`PushScopedSymbolNode`] nodes,
+//!     which push onto the symbol stack new things for us to look for
+//!   - [_pop symbol_][`PopSymbolNode`] and [_pop scoped symbol_][`PopScopedSymbolNode`] nodes,
+//!     which pop things off the symbol stack once they've been found
+//!   - [_drop scopes_][`DropScopesNode`] and [_jump to scope_][`JumpToNode`] nodes, which
+//!     manipulate the scope stack
+//!
+//! [`DropScopesNode`]: struct.DropScopesNode.html
+//! [`ExportedScopeNode`]: struct.ExportedScopeNode.html
+//! [`InternalScopeNode`]: struct.InternalScopeNode.html
+//! [`JumpToNode`]: struct.JumpToNode.html
+//! [`PushScopedSymbolNode`]: struct.PushScopedSymbolNode.html
+//! [`PushSymbolNode`]: struct.PushSymbolNode.html
+//! [`PopScopedSymbolNode`]: struct.PopScopedSymbolNode.html
+//! [`PopSymbolNode`]: struct.PopSymbolNode.html
+//! [`RootNode`]: struct.RootNode.html
+//!
+//! All nodes except for the singleton _root node_ and _jump to scope_ node belong to
```

Comment:
That's a good question!  Drop scopes _can't_ be a singleton, because there are nodes coming out of it, and different _drop scopes_ nodes will have different outgoing edges.  If we collapsed all _drop scopes_ nodes into a singleton we'd end up merging together all of those outgoing edges that would break the semantic of the path-finding algorithm.